### ISSUE TITLE
Add support for operations on List types to cls_tabular class

### ISSUE THREAD

**Issue by jlefevre** [uccross/skyhookdm-ceph-cls#38]
[Awkward arrays](https://github.com/scikit-hep/awkward-array) are currently stored as [lists within Arrow tables](https://arrow.apache.org/docs/cpp/api/datatype.html#classarrow_1_1_list_type) inside Skyhook.  This project will investigate and implement a small subset of `awkward array` (or list) operations that can be offloaded ("pushed down") into storage for query processing. Common list manipulations that perform data reduction such as filters or summary/agg methods will be most useful to apply within storage, since these will reduce network IO transferred back to the client from the storage layer.

This will involve the following steps.

We have jagged arrays types defined here, but do not yet support predicate operations on this data type: [link](https://github.com/uccross/skyhookdm-ceph-cls/blob/luminous/src/cls/cls_tabular_utils.h#L110)

Apply predicates will be called here, which will include the specified new ops: [link](https://github.com/uccross/skyhookdm-ceph-cls/blob/luminous/src/cls/cls_tabular_processing.cc#L384)

Add a `case` to apply predicates on lists/jaggedarrays type: [link](https://github.com/uccross/skyhookdm-ceph-cls/blob/luminous/src/cls/cls_tabular_utils.cc#L1777)

Add a compare (or similar) function, that applies the new ops on lists.  Here is an example of a simple integer compare op: [link](https://github.com/uccross/skyhookdm-ceph-cls/blob/luminous/src/cls/cls_tabular_utils.cc#L2030).  And a regex op and date op:  [link](https://github.com/uccross/skyhookdm-ceph-cls/blob/luminous/src/cls/cls_tabular_utils.cc#L2083)

Define any new operation types for list, here: [link](https://github.com/uccross/skyhookdm-ceph-cls/blob/luminous/src/cls/cls_tabular_utils.h#L131)

Adding support for operations on list types will be useful in general in Skyhook for any supported data format.

**Comment by carlosmalt** [comment]
Hi Jeff, Can you provide a bit more info on _why_ instead of what? Who wants this functionality? I also like to see a plan of how this work can be isolated from any effects due to `awkward` rewrites (one already happened since the original release). Also, it would be so nice if the debugging work that Jim and others already performed on equivalent functionality within`awkward` doesn't have to be repeated here. Perhaps this is a good use case for thinking about how to run libraries within Ceph/CLS using outcomes of Saloni's work on WebAssembly isolates.

**Comment by jlefevre** [comment]
Good points Carlos, thanks.  I totally agree and updated the beginning of the comment to add motivation.  We should definitely not duplicate awkward functionality, and get in the business of closely tracking awkward lib.  It was really meant to use as a motivation of the types of list ops scientific analysts use.  Probably the simple ops can apply to lists in general so I think this will be very useful.  Regarding WebAssembly, yes I think once we understand how to express and offload list predicates, we should definitely consider the feasibility of a WebAssembly approach for awkward.

**Comment by carlosmalt** [comment]
Now that I know a bit more about Arrow, doesn't this duplicate Arrow `compute` and `function` functionality?

**Comment by jlefevre** [comment]
Yes we can use the array ops available in Arrow.  Xiongfeng is working on this.
